src/controller/proxy/local.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/goharbor/harbor/src/lib"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/pkg/registry"
+	"io"
+	"time"
+)
+
+// localInterface defines operations related to local repo under proxy mode
+type localInterface interface {
+	// BlobExist check if the blob exist in local repo
+	BlobExist(ctx context.Context, art lib.ArtifactInfo) (bool, error)
+	// Manifest check if the manifest exist in local repo
+	ManifestExist(ctx context.Context, art lib.ArtifactInfo) (bool, error)
+	// PushBlob push blob to local repo
+	PushBlob(localRepo string, desc distribution.Descriptor, bReader io.ReadCloser) error
+	// PushManifest push manifest to local repo
+	PushManifest(repo string, tag string, manifest distribution.Manifest) error
+	// PushManifestList push manifest list to local repo
+	PushManifestList(ctx context.Context, repo string, tag string, man distribution.Manifest) error
+	// CheckDependencies check if the manifest's dependency is ready
+	CheckDependencies(ctx context.Context, repo string, man distribution.Manifest) []distribution.Descriptor
+	// DeleteManifest cleanup delete tag from local cache
+	DeleteManifest(repo, ref string)
+}
+
+func (l *localHelper) ManifestExist(ctx context.Context, art lib.ArtifactInfo) (bool, error) {
+	a, err := l.artifactCtl.GetByReference(ctx, art.Repository, art.Digest, nil)
+	if err != nil {
+		return false, nil
+	}
+	return a != nil, nil
+}
+
+// localHelper defines operations related to local repo under proxy mode
+type localHelper struct {
+	registry    registry.Client
+	artifactCtl artifactController
+}
+
+type artifactController interface {
+	GetByReference(ctx context.Context, repository, reference string, option *artifact.Option) (artifact *artifact.Artifact, err error)
+}
+
+// newLocalHelper create the localInterface
+func newLocalHelper() localInterface {
+	l := &localHelper{artifactCtl: artifact.Ctl}
+	l.init()
+	return l
+}
+
+func (l *localHelper) BlobExist(ctx context.Context, art lib.ArtifactInfo) (bool, error) {
+	return l.registry.BlobExist(art.Repository, art.Digest)
+}
+
+func (l *localHelper) init() {
+	if l.registry != nil {
+		return
+	}
+	log.Debugf("core url:%s, local core url: %v", config.GetCoreURL(), config.LocalCoreURL())
+	// the traffic is internal only
+	registryURL := config.LocalCoreURL()
+	authorizer := comHttpAuth.NewSecretAuthorizer(config.ProxyServiceSecret)
+	l.registry = registry.NewClientWithAuthorizer(registryURL, authorizer, true)
+}
+
+func (l *localHelper) PushBlob(localRepo string, desc distribution.Descriptor, bReader io.ReadCloser) error {
+	log.Debugf("Put blob to local registry, localRepo:%v, digest: %v", localRepo, desc.Digest)
+	err := l.registry.PushBlob(localRepo, string(desc.Digest), desc.Size, bReader)
+	return err
+}
+
+func (l *localHelper) PushManifest(repo string, tag string, manifest distribution.Manifest) error {
+	// Make sure there is only one go routing to push current artName to local repo
+	if len(tag) == 0 {
```

Comment:
what if request is to pull manifest by digest?

Reply:
it will be pushed and tagged with latest